### PR TITLE
Add a ServerType for peons

### PR DESCRIPTION
**Pull request by jon-wei**
I encountered a failure in a kafka indexing task after #4148 introduced ServerType:

```
Caused by: java.lang.IllegalArgumentException: No enum constant io.druid.server.coordination.ServerType.INDEXER-EXECUTOR
	at java.lang.Enum.valueOf(Enum.java:238)
	at io.druid.server.coordination.ServerType.valueOf(ServerType.java:22)
	at io.druid.server.coordination.ServerType.fromString(ServerType.java:58)
	at io.druid.server.coordination.DruidServerMetadata.<init>(DruidServerMetadata.java:50)
	at io.druid.guice.StorageNodeModule.getMetadata(StorageNodeModule.java:71)
	at io.druid.guice.StorageNodeModule$$FastClassByGuice$$a8289f44.invoke(<generated>)
...
	at io.druid.guice.LifecycleModule$2.start(LifecycleModule.java:154)
	at io.druid.cli.GuiceRunnable.initLifecycle(GuiceRunnable.java:103)
	at io.druid.cli.CliPeon.run(CliPeon.java:279)
	at io.druid.cli.Main.main(Main.java:109)
```

This PR adds a ServerType for peons which have a nodeType of "indexer-executor".

### CODE REVIEW

**Comment by jihoonson** [comment]
@jon-wei thanks. I missed this server type. The code looks good overall. Would you address @leventov's comment?

**Comment by jon-wei** [comment]
Thanks for the review!

@leventov 
Re: the enum conversion, good catch on the back conversions, though I'm not sure what the best approach is here, I could override toString() for the peon type such that it returns "indexer-executor" or do the single-word change.

I initially considered renaming the nodeType to one word or switching the hyphen to an underscore, but I'm not familiar with the Druid-Zookeeper interactions, and wasn't clear on whether the rename would break something related to node/path handling there.

I'm also wary of changing it for external compatibility reasons, I see at least one external piece of code that appears to assume that nodeType name, not sure if there are other users out there that rely on the "indexer-executor" string:
https://github.com/druid-io/druid-console/blob/master/src/client/factories/hUtils.coffee#L25

If someone has a clearer view on the impact of changing the nodeType string, feedback would be much appreciated.

---------

Re: the switch in `DruidCluster`, throwing an error in the switch for the peon type is fine since the peons are not intended to be segment replication targets.

**Comment by jihoonson** [comment]
> Re: the switch in DruidCluster, throwing an error in the switch for the peon type is fine since the peons are not intended to be segment replication targets.

Throwing an exception looks fine, but it would be better to improve the error message. Currently it simply complains about ```unknown server type```, but I think it should be ```unsupported server type```.

**Comment by jon-wei** [comment]
@leventov  @jihoonson 

Updated the patch:

- Added a case to the switch statement for INDEXER_EXECUTOR that throws an exception with "Unsupported server type"

- Overrode toString() method for ServerType to lowercase and convert underscores to hyphens, ended up deciding to keep the old format for backwards compatibility

- Adds a fix to ServersResource so that ServerType.toString() is used for the "type" field instead of the enum object itself, this previously broke the coordinator console's server view

File: server/src/main/java/io/druid/server/coordination/ServerType.java [hunk 1]
```diff
@@ -19,10 +19,30 @@
 
 package io.druid.server.coordination;
 
+/**
+ * This enum represents types of druid services that hold segments.
```
Comment (jihoonson) [inline server/src/main/java/io/druid/server/coordination/ServerType.java]: This might be not valid because ```INDEXER_EXECUTOR``` cannot hold segments?

Comment (jon-wei) [inline server/src/main/java/io/druid/server/coordination/ServerType.java]: The peons will hold segments for some time as they're indexing, before they're handed off

Comment (jihoonson) [inline server/src/main/java/io/druid/server/coordination/ServerType.java]: Ah, it makes sense.

File: server/src/main/java/io/druid/server/coordination/ServerType.java [hunk 2]
```diff
@@ -19,6 +19,18 @@
...
+ * These types are externally visible (e.g., from the output of /druid/coordinator/v1/servers).
+ *
+ * For backwards compatibility, when presenting these types externally, the toString() representation
+ * of the enum should be used.
+ *
+ * The toString() method converts the enum name() to lowercase and replaces underscores with hyphens,
+ * which is the format expected for the server type string prior to the patch that introduced ServerType:
+ * https://github.com/druid-io/druid/pull/4148
+ */
 public enum ServerType
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordination/ServerType.java]: Shouldn't you annotate `fromString()` as `@JsonCreator`? Or it works already?

Comment (jon-wei) [inline server/src/main/java/io/druid/server/coordination/ServerType.java]: I don't think there are instances where a ServerType is being deserialized directly, DruidServer and DruidServerMetadata would be such locations but they currently take a String "type" and DruidServerMetadata converts that String to a ServerType with fromString().

I'll look into adding @JsonCreator here and changing DruidServer/DruidServerMetadata to accept a ServerType parameter.

**Comment by pjain1** [comment]
This looks good to me. @jon-wei are you planning to add anything more in this for @leventov's comment ?

**Comment by jon-wei** [comment]
@pjain1 Thanks, I'm going to push another update that fixes some unit test failures and the String->ServerType changes I mentioned

**Comment by pjain1** [comment]
@jon-wei thanks looking forward to it, lets get this fixed soon.

**Comment by jon-wei** [comment]
Added @JsonCreator and @JsonValue to fromString()/toString() in ServerType, changed DruidServer and DruidServerMetadata to accept a ServerType parameter instead of a type string

**Pull request merged**